### PR TITLE
Initial commit with one class and one example.

### PR DESCRIPTION
**Pull request by hertschuh**
- Added `BruteForceRetrieval` layer
- Added basic retrieval example
- Added standard shell scripts to format, lint, and generate the api with namex
- Added config for black, isort, flake8 and mypy

Unit tests and example were run with all 3 backends (tf, jax and torch).

### CODE REVIEW

File: keras_rs/src/types.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+from typing import Any, Optional, Sequence
+
+"""
+A tensor in any of the backends.
+
+We do not define it explicitly to not require all the backends to be installed
+and imported. The explicit definition would be:
+```
+Union[
```
Comment (fchollet) [inline keras_rs/src/types.py:11]: Might also include NumPy array (for numpy backend + fact that layers will accept and convert any numpy array)

File: keras_rs/src/layers/retrieval/brute_force_retrieval.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    identifiers, but it can be configured to return a single tensor with the top
+    identifiers.
+
+    The identifiers for the candidates can be specified as a tensor. If not
+    provided, the IDs used are simply the candidate indices.
+    """
+
+    def __init__(
+        self,
+        candidate_embeddings: Optional[types.Tensor] = None,
```
Comment (fchollet) [inline keras_rs/src/layers/retrieval/brute_force_retrieval.py:28]: Do we actually need type annotations? They aren't in use in any other package. My take is that we should skip them for non-standard types (e.g. you can annotate int or string or bool but not Tensor).

File: keras_rs/src/layers/retrieval/brute_force_retrieval.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        candidate_ids: Optional[types.Tensor] = None,
+        k: int = 10,
+        return_scores: bool = True,
+        name: Optional[Text] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Initializes the layer.
+
+        Args:
```
Comment (fchollet) [inline keras_rs/src/layers/retrieval/brute_force_retrieval.py:37]: This section should be in the class docstring.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras_rs/src/layers/retrieval/brute_force_retrieval.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+import keras
+
+from keras_rs.src import types
+from keras_rs.src.api_export import keras_rs_export
+
+
+@keras_rs_export("keras_rs.layers.BruteForceRetrieval")
+class BruteForceRetrieval(keras.layers.Layer):
+    """Brute force top-k retrieval.
```
[the remark this answers is not in the document]
Comment (hertschuh) [reply keras_rs/src/layers/retrieval/brute_force_retrieval.py:11]: Added, although it's pretty minimal.

File: keras_rs/src/layers/retrieval/brute_force_retrieval.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    top-k candidates for a given query. It does this by computing the scores for
+    all of the candidates for the query and extracting the top ones. The
+    returned top-k candidates are sorted by score.
+
+    By default, this layer returns a tuple with the top scores and top
+    identifiers, but it can be configured to return a single tensor with the top
+    identifiers.
+
+    The identifiers for the candidates can be specified as a tensor. If not
+    provided, the IDs used are simply the candidate indices.
```
[the remark this answers is not in the document]
Comment (hertschuh) [reply keras_rs/src/layers/retrieval/brute_force_retrieval.py:23]: Moved from below.

File: keras_rs/src/types.py (hunk 1 above)
Comment (hertschuh) [reply keras_rs/src/types.py:11]: Added.

File: keras_rs/src/layers/retrieval/brute_force_retrieval.py (hunk 3 above)
Comment (hertschuh) [reply keras_rs/src/layers/retrieval/brute_force_retrieval.py]: Moved up.

File: keras_rs/src/layers/retrieval/brute_force_retrieval.py (hunk 2 above)
Comment (hertschuh) [reply keras_rs/src/layers/retrieval/brute_force_retrieval.py:54]: That's a good question. This is a bit of an experiment that I took from the internal prototype. My findings so far are that:

- I would be doable because Keras RS should be fairly simple and shouldn't have many cases of arguments that can be one of many different types.
- I personally like annotating tensors to make it clear when it's a single tensor vs. a tuple of tensors vs. an arbitrary structure.
- Type annotation is an all-or-nothing deal. It doesn't really work to partially annotate the code. One main reason is that you cannot do any basic verification (I'm using `mypy`) to verify that they're not completely wrong (which is worse than no annotation).
- The fact that Keras is not annotated (which would be really hard) diminishes the value of annotating here.

Bottom line, I can remove if you prefer.

**Review by fchollet — APPROVED** [review]
